content/en/docs/components/pipelines/author-a-pipeline/components.md

```diff
@@ -164,7 +164,77 @@ To create a containerized component, you must:
 
 ### 3. Custom container components
-<!-- @zichuan-scott-xu to provide this section -->
+
+Container-based components make it easier to use the pre-existed container to build a KFP component. 
+The user will control the container entrypoint and is able to define the component by specifying their container, commands, and args in a `ContainerSpec` object with similar authoring experiences as defining Python components. 
+
+To define a container component, you must:
+1) Write your component’s code as a python function that returns a `dsl.ContainerSpec` object to specify the container image and the commands to be run in the container and wrap the function into a `@container_component` decorator. The function should satisfy the following requirements:
+
+    *   It should not use any code declared outside of the function definition.
+    *   The function should do nothing other than returning a `dsl.ContainerSpec`
+        object, with the following parameters:
+        * image: The target image where the command (with args) is executed. The user will control the entrypoint of the image.
+        * command(optional): The command to be executed. 
+        * args(optional): The arguments of the command. It’s recommended to place the input of the     components in the args section instead of the command section.
+
+    The decorator will then compose a `ComponentSpec` object using the `ContainerSpec` to be compiled for execution. (Learn more about [ContainerSpec][dsl-reference-documentation] in documentation.) 
+
+2) Specify your function's inputs and outputs in the function's    signature [Learn more about passing data between components][data-passing]. Specifically for container-based component, your function's inputs and outputs must meet the following requirements:
+    
+    *   All your function's arguments must have data type annotations.
+    *   Different from a Python component, your return type annotation for the function 
+        must either be `dsl.ContainerSpec` or omitted. 
+    *   If the function accepts or returns large amounts of data or complex
+        data types, you must annotate that argument as an _artifact_. Note that in the function you defined, you can only access artifacts via its `.url`, `.path`, or `.metadata` attribute. Accessing any other attribute or the artifact variable by itself is currently not allowed. 
+
+Here is a minimal example that uses a bash script to demonstrate the use of container component:
+```python
+from kfp import dsl
+@dsl.container_component
+def hello_world_comp(text: str, out_filename: dsl.OutputPath(str)):
+    return dsl.ContainerSpec(
+        image='python:3.7' # or put your custom container image here
+        command=['source', 'hello_world.sh'],
+        args=['--in', text, '--out', out_filename]
+    )
+```
+
+Meanwhile, add a bash script `hello_world.sh` on your container that takes outputs the text into the provided path:
```

Comment:
This is a tricky example, since `python:3.7` is a pre-existing image that will not include `hello_world.sh` and there is no way to inject it into the container. Additionally, I think this example would fail, because the `$4` file does not exist until you create it.

What do you think about removing this example, then using the pipeline below with a bit more detail on each component?

Reply:
I guess it's actually okay to use `>>` for piping to create a new file if the file doesn't exist yet. But I agree that the pipeline below would be more suitable in this case.